File: src/tree_matching/bipartite_graph.rs
use super::node::Node;
use std::collections::HashMap;

#[derive(PartialEq, Debug)]
pub struct Edge {
    pub source: Node,
    pub target: Node,
    pub value : f64,
}

#[derive(PartialEq, Debug)]
pub struct BipartiteGraph {
    pub node_group_1: Vec<Node>,
    pub node_group_2: Vec<Node>,
    pub edges: Vec<Edge>,
}

impl Clone for Edge {
    fn clone(&self) -> Self {
        Edge {
            source: self.source.clone(), // Cloning the inner Node
            target: self.target.clone(), // Cloning the inner Node
            value: self.value,
        }
    }
}

impl Edge {
    pub fn new(source: Node, target: Node) -> Edge {
        Edge {
            source,
            target,
            value: 1.0,
        }
    }
}

impl BipartiteGraph {

    pub fn new(node_group_1: Vec<Node>, node_group_2: Vec<Node>, edges: Vec<Edge>) -> BipartiteGraph {
        BipartiteGraph {
            node_group_1,
            node_group_2,
            edges,
        }
    }

    pub fn find_node(&self, group: usize, name: &str) -> Option<&Node> {
        let node_group = match group {
            1 => &self.node_group_1,
            2 => &self.node_group_2,
            _ => return None,
        };

        for node in node_group {
            if node.name == name {
                return Some(node);
            }
        }

        None
    }

    pub fn get_group(&self, name: &str) -> Option<usize> {
        if self.find_node(1, name).is_some() {
            return Some(1);
        }
        if self.find_node(2, name).is_some() {
            return Some(2);
        }
        None
    }

    pub fn add_node(&mut self, group: usize, node: Node) {
        match group {
            1 => self.node_group_1.push(node),
            2 => self.node_group_2.push(node),
            _ => (),
        }
    }

    pub fn remove_node(&mut self, node: &Node) {
        self.node_group_1.retain(|n| n != node);
        self.node_group_2.retain(|n| n != node);
    }

    pub fn add_edge(&mut self, edge: Edge) -> Result<(), &'static str> {
        if self.get_group(&edge.source.name) == self.get_group(&edge.target.name) {
            return Err("Both nodes are in the same group");
        }
        self.edges.push(edge);
        Ok(())
    }

    pub fn remove_edge(&mut self, edge: &Edge) {
        self.edges.retain(|e| e != edge);
    }

    pub fn remove_all_adjacent_edges(&mut self, edge: &Edge) {
        self.edges.retain(|e| !(e.source == edge.source) ^ (e.target == edge.target));
    }

    pub fn build_edges_from_neighborhoods(&mut self, neighborhoods: &HashMap<&Node, HashMap<&Node, f64>>) {
        for (node, neighborhood) in neighborhoods {
            self.build_edges_from_neighborhood(node, neighborhood);
        }
    }

    fn build_edges_from_neighborhood(&mut self, node: &Node, neighborhood: &HashMap<&Node, f64>) {
        for (&node_in_neighborhood, &value) in neighborhood {
            self.edges.push(Edge { source: node.clone(), target: node_in_neighborhood.clone(), value:value});
        }
    }
}

File: src/main/java/bipartiteGraph/BipartiteGraph.java
package bipartiteGraph;

import fileTree.FileTree;
import inMemory.AttributeSetter;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Collections;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ConcurrentHashMap;

/**
 * This class represents a bipartite graph, that is used during the matching process.
 */
@Getter
public class BipartiteGraph {
    @Setter
    private List<Node> nodeGroup1;
    @Setter
    private List<Node> nodeGroup2;
    @Setter
    private List<Edge> edges;
    private final AttributeSetter attributeSetter = new AttributeSetter();

    public static BipartiteGraph buildFromTrees(FileTree tree1, FileTree tree2) {
        BipartiteGraph graph = new BipartiteGraph();
        List<Node> nodes1 = new ArrayList<>();
        List<Node> nodes2 = new ArrayList<>();
        graph.initNodes(tree1, nodes1, 1);
        graph.initNodes(tree2, nodes2, 2);
        graph.addNodes(nodes1, 1);
        graph.addNodes(nodes2, 2);
        return graph;
    }

    public BipartiteGraph() {
        this.nodeGroup1 = new ArrayList<>();
        this.nodeGroup2 = new ArrayList<>();
        this.edges = new ArrayList<>();
    }

    /**
     * Adds a list of nodes to the graph.
     * @param nodes the list of nodes
     * @param group the group of the nodes
     */
    public void addNodes(List<Node> nodes, int group) {
        for (Node node : nodes) {
            addNode(node, group);
        }
    }

    /**
     * Finds a node in a group.
     * @param group the group
     * @param node the node
     * @return the node if it is found, null otherwise
     */
    public Node findNode(int group, Node node) {
        List<Node> nodeGroup = (group == 1) ? nodeGroup1 : (group == 2) ? nodeGroup2 : null;
        if (nodeGroup == null) return null;

        for (Node n : nodeGroup) {
            if (n == node) {
                return node;
            }
        }
        return null;
    }

    /**
     * Gets the group of a node.
     * @param node the node
     * @return the group of the node
     */
    public Integer getGroup(Node node) {
        if (findNode(1, node) != null) {
            return 1;
        }
        if (findNode(2, node) != null) {
            return 2;
        }
        return null;
    }

    /**
     * Adds a node to the graph.
     * @param node the node
     * @param group the group of the node
     */
    public void addNode(Node node, int group) {
        if (group == 1) {
            nodeGroup1.add(node);
        } else if (group == 2) {
            nodeGroup2.add(node);
        }
    }

    /**
     * Adds a list of edges to the graph.
     * @param edges the list of edges
     * @throws IllegalArgumentException if both nodes are in the same group
     */
    public void addEdges(List<Edge> edges) throws IllegalArgumentException {
        for (Edge edge : edges) {
            addEdge(edge);
        }
    }

    /**
     * Adds an edge to the graph.
     * @param edge the edge
     * @throws IllegalArgumentException if both nodes are in the same group
     */
    public void addEdge(Edge edge) throws IllegalArgumentException {
        Integer sourceGroup = getGroup(edge.getSource());
        Integer targetGroup = getGroup(edge.getTarget());

        if (sourceGroup == null || targetGroup == null || sourceGroup.equals(targetGroup)) {
            throw new IllegalArgumentException("Both nodes are in the same group");
        }

        this.edges.add(edge);
    }

    /**
     * Removes an edge from the graph.
     * @param edge the edge
     */
    public void removeEdge(Edge edge) {
        this.edges.remove(edge);
    }

    /**
     * Builds the edges of the graph from a map of neighborhoods.
     * @param neighborhoods the map of neighborhoods
     */
    public void buildEdgesFromNeighborhoods(Map<Node, HashMap<Node, Double>> neighborhoods) {
        for (Map.Entry<Node, HashMap<Node, Double>> entry : neighborhoods.entrySet()) {
            Node node = entry.getKey();
            Map<Node, Double> neighborhood = entry.getValue();
            buildEdgesFromNeighborhood(node, neighborhood);
        }
    }

    /**
     * Builds the edges of the graph from a neighborhood.
     * @param node the node
     * @param neighborhood the neighborhood
     */
    private void buildEdgesFromNeighborhood(Node node, Map<Node, Double> neighborhood) {
        for (Map.Entry<Node, Double> entry : neighborhood.entrySet()) {
            Node neighbor = entry.getKey();
            double value = entry.getValue();
            if (value > 0.0) edges.add(new Edge(node, neighbor, value));
        }
    }

    @Override
    public String toString() {
        return "BipartiteGraph{" +
                "nodeGroup1=" + nodeGroup1 +
                ", nodeGroup2=" + nodeGroup2 +
                ", edges=" + edges +
                '}';
    }

    private void initNodes(FileTree tree, List<Node> nodes, int group) {
        ConcurrentMap<Node, Integer> parents = new ConcurrentHashMap<>();
        List<FileTree> fileTreeNodes = tree.getNodes();
        List<Node> synchronizedNodes = Collections.synchronizedList(new ArrayList<>());
        createNodes(fileTreeNodes, synchronizedNodes, parents, group);
        nodes.addAll(synchronizedNodes);
        setParentRelationships(parents, synchronizedNodes);
    }

    private void createNodes(List<FileTree> fileTreeNodes, List<Node> synchronizedNodes, ConcurrentMap<Node,Integer> parents, int group) {
        fileTreeNodes.parallelStream().forEach(fileTree -> {
            Node node;
            if (fileTree.getGraphNode() == null) {
                node = new Node(fileTree.getPath());
                node.setId(fileTree.getId());
                node.setGroup(group);
                attributeSetter.tokenize(node);
                synchronizedNodes.add(node);
                fileTree.setGraphNode(node);
            } else {
                node = fileTree.getGraphNode();
            }
            if (fileTree.getParent() != null) {
                parents.put(node, fileTree.getParent().getId());
            }
        });
    }

    private void setParentRelationships(ConcurrentMap<Node, Integer> parents, List<Node> synchronizedNodes) {
        parents.entrySet().parallelStream().forEach(entry -> {
            Node node = entry.getKey();
            Integer parentId = entry.getValue();
            Node parent = synchronizedNodes.stream().filter(n -> n.getId() == parentId).findFirst().orElse(null);
            node.setParent(parent);
            if (parent != null) {
                synchronized (parent) {
                    parent.getChildren().add(node);
                }
            }
        });
    }

}